src/services/refactors/convertImport.ts

```diff
@@ -0,0 +1,104 @@
+/* @internal */
+namespace ts.refactor.generateGetAccessorAndSetAccessor {
+    const refactorName = "Convert import";
+    const actionNameNamespaceToNamed = "Convert namespace import to named imports";
+    const actionNameNamedToNamespace = "Convert named imports to namespace import";
+    registerRefactor(refactorName, {
+        getAvailableActions(context): ApplicableRefactorInfo[] | undefined {
+            const i = getImportToConvert(context);
+            if (!i) return undefined;
+            const description = i.kind === SyntaxKind.NamespaceImport ? Diagnostics.Convert_namespace_import_to_named_imports.message : Diagnostics.Convert_named_imports_to_namespace_import.message;
+            const actionName = i.kind === SyntaxKind.NamespaceImport ? actionNameNamespaceToNamed : actionNameNamedToNamespace;
+            return [{ name: refactorName, description, actions: [{ name: actionName, description }] }];
+        },
+        getEditsForAction(context, actionName): RefactorEditInfo {
+            Debug.assert(actionName === actionNameNamespaceToNamed || actionName === actionNameNamedToNamespace);
+            const edits = textChanges.ChangeTracker.with(context, t => doChange(context.file, context.program, t, Debug.assertDefined(getImportToConvert(context))));
+            return { edits, renameFilename: undefined, renameLocation: undefined };
+        }
+    });
+
+    // Can convert imports of the form `import * as m from "m";` or `import d, { x, y } from "m";`.
+    function getImportToConvert(context: RefactorContext): NamedImportBindings | undefined {
+        const { file } = context;
+        const span = getRefactorContextSpan(context);
+        const token = getTokenAtPosition(file, span.start, /*includeJsDocComment*/ false);
+        const importDecl = getParentNodeInSpan(token, file, span);
+        if (!importDecl || !isImportDeclaration(importDecl)) return undefined;
+        const { importClause } = importDecl;
+        return importClause && importClause.namedBindings;
+    }
+
+    function doChange(sourceFile: SourceFile, program: Program, changes: textChanges.ChangeTracker, toConvert: NamedImportBindings): void {
+        const usedIdentifiers = createMap<true>();
+        forEachFreeIdentifier(sourceFile, id => usedIdentifiers.set(id.text, true));

```

Comment:
Looks like `isFileLevelUniqueName` isn't suitable for this purpose since `sourceFile.identifiers` includes the text of every `Identifier` node in the source file, including the names in property accesses. So if we access `import * as m from "m"; m.x` then that would make us convert it to `import { x as _ x } from "m";` unnecessarily.

Reply:
D'oh.. did not think of this one.. i guess we need a walk after all.

Reply:
one other option is to call resolveName for the property name on every reference for the namespace import, if it resolves to anything then you know you need to alias it.